whisker-backend/pkg/handlers/v1/flows.go

```diff
@@ -0,0 +1,114 @@
+package v1
+
+import (
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/projectcalico/calico/goldmane/pkg/client"
+	"github.com/projectcalico/calico/goldmane/proto"
+	apictx "github.com/projectcalico/calico/lib/httpapimachinery/pkg/context"
+	"github.com/projectcalico/calico/lib/httpapimachinery/pkg/handler"
+	whiskerv1 "github.com/projectcalico/calico/whisker-backend/pkg/apis/v1"
+)
+
+type flowsHdlr struct {
+	flowCli client.FlowRetrieverClient
+}
+
+func NewFlows(cli client.FlowRetrieverClient) *flowsHdlr {
+	return &flowsHdlr{cli}
+}
+
+func (hdlr *flowsHdlr) APIs() []handler.API {
+	return []handler.API{
+		{
+			Method:  http.MethodGet,
+			URL:     whiskerv1.FlowsPath,
+			Handler: handler.NewJSONListResponseHandler(hdlr.List),
+		},
+		{
+			Method:  http.MethodGet,
+			URL:     whiskerv1.FlowsStreamPath,
+			Handler: handler.NewJSONEventStreamHandler(hdlr.Stream),
+		},
+	}
+}
+
+func (hdlr *flowsHdlr) List(ctx apictx.Context, params whiskerv1.ListFlowsParams) handler.ListResponse[whiskerv1.FlowResponse] {
+	logger := ctx.Logger()
+	logger.Debug("List flows called.")
+
+	flowReq := &proto.FlowRequest{}
+	if !params.StartTimeGt.IsZero() {
+		flowReq.StartTimeGt = params.StartTimeGt.Unix()
+	}
+
+	if !params.StartTimeLt.IsZero() {
+		flowReq.StartTimeLt = params.StartTimeLt.Unix()
+	}
+
+	flows, err := hdlr.flowCli.List(ctx, flowReq)
+	if err != nil {
+		logger.WithError(err).Error("failed to list flows")
+		return handler.NewListResponse[whiskerv1.FlowResponse](500).SetErrorMsg("Internal Server Error")
+	}
+
+	var rspFlows []whiskerv1.FlowResponse
+	for _, flow := range flows {
+		rspFlows = append(rspFlows, protoToFlow(flow))
+	}
+
+	return handler.NewListResponse[whiskerv1.FlowResponse](http.StatusOK).SetItems(rspFlows)
+}
+
+func (hdlr *flowsHdlr) Stream(ctx apictx.Context, params whiskerv1.StreamFlowsParams, rspStream handler.EventStream[whiskerv1.FlowResponse]) {
+	logger := ctx.Logger()
+	logger.Debug("Stream flows called.")
+
+	// TODO figure out how we're going to handle errors.
+	flowStream, err := hdlr.flowCli.Stream(ctx, &proto.FlowRequest{})
+	if err != nil {
+		logger.WithError(err).Error("failed to stream flows")
+		return
+	}
+
+	for {
+		flow, err := flowStream.Recv()
+		if err != nil {
+			logger.WithError(err).Error("failed to stream flows")
+			break
+		}
+
+		if err := rspStream.Data(protoToFlow(flow)); err != nil {
+			logger.WithError(err).Error("failed to write flow response")
+			return
+		}
+	}
+
+	return
+}
+
+func protoToFlow(flow *proto.Flow) whiskerv1.FlowResponse {
+	return whiskerv1.FlowResponse{
+		StartTime: time.Unix(flow.StartTime, 0),
+		EndTime:   time.Unix(flow.EndTime, 0),
+		Action:    flow.Key.Action,
+
+		SourceName:      flow.Key.SourceName,
+		SourceNamespace: flow.Key.SourceNamespace,
+		SourceLabels:    strings.Join(flow.SourceLabels, " | "),
```

Comment:
Is a single string of Labels really what the UI wants instead of a json list?

Reply:
No I wouldn't think so, this was just conforming to what Ronan already had.